### PR TITLE
feat: Print version with Client and Server

### PR DESCRIPTION
**Pull request by ahsan-z-khan**
Issue #, if available:
#12 

*Description of changes:*

Update the implementation `finch version` according to https://github.com/runfinch/finch/issues/12#issuecomment-1347252973

*Testing done:*
Yet to be updated

- [x] I've reviewed the guidance in CONTRIBUTING.md

#### License Acceptance

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by davidhsingyuchen — COMMENTED** [review]
Leaving a comment now although it's still a draft PR because we want to make the discussion  transparent:

> my question is, what other way we can call nerdctl from version commant to generate finch and nerdctl version together

If `--format` is not there, we can probably just print finch version followed by printing the thing returned from `nerdctl version` so that we don't need to duplicate [the printing logic](https://github.com/containerd/nerdctl/blob/854985261da22f0e435ad963ecb13fbbaa37758a/cmd/nerdctl/version.go#L73-L94). That way, if a new thing is added to the output of `nerdctl version`, we don't have to update our code.

If `--format` is there, we may need to mimic [`VersionInfo`](https://github.com/containerd/nerdctl/blob/854985261da22f0e435ad963ecb13fbbaa37758a/pkg/inspecttypes/dockercompat/info.go#L72) to define a struct like:

```go
type VersionInfo struct {
	Finch FinchVersion
	Client ClientVersion
	Server *ServerVersion
}
```

Add then we can

1. Call `nerdctl version --format json` to populate the nerdctl part of the struct.
2. Populate the finch part of the struct.
3. Handle the template ([example](https://github.com/containerd/nerdctl/blob/854985261da22f0e435ad963ecb13fbbaa37758a/cmd/nerdctl/version.go#L55-L71)).

Please let me know if this makes sense to you, thanks!

File: cmd/finch/version.go
Comment (pendo324) [inline cmd/finch/version.go:48]: I think you can just use the `va.creator` again, instead of shelling out with exec:

```suggestion
        limaArgs := []string{"shell", limaInstanceName, "nerdctl", "version", "--format", "json"}
	out, err := va.creator.CreateWithoutStdio(limaArgs...).Output()
```

**Comment by ningziwen** [comment]
I added one comment to the issue. I think we should close the final output format first before reviewing the PR.

File: cmd/finch/version.go
Comment (ChandruShetty) [inline cmd/finch/version.go:42]: Suggest adding lima version too.

**Review by ChandruShetty — DISMISSED** [review]

File: cmd/finch/version.go
Comment (azhouwd) [inline cmd/finch/version.go:77]: If the VM is in a broken status, we print `Finch version: ...`. Should we also print the `Finch version: ...` if the VM is in good status? Also should we print out some warning message to inform users that the output is printed due to VM is broken?

File: cmd/finch/version.go
Comment (azhouwd) [inline cmd/finch/version.go:81]: It's better to do the error handling here:
```suggestion
	out, err := va.creator.CreateWithoutStdio(limaArgs...).Output()
	if err != nil {...}
```

File: cmd/finch/version.go
Comment (azhouwd) [inline cmd/finch/version.go:90]: I assume we can directly call `fmt.Printf("Version:\t%s\n", version.Version)` here, curious about what was the intention to use `Fprintf`?

File: cmd/finch/version_test.go
Comment (azhouwd) [inline cmd/finch/version_test.go:45]: For the version info related to nerdctl, we can mock the result from `CreateWithoutStdio`, an example to follow:
https://github.com/runfinch/finch/blob/main/cmd/finch/virtual_machine_init_test.go#L239
from line 239 to line 241.

File: e2e/version_test.go
Comment (azhouwd) [inline e2e/version_test.go:17]: The only way I can come up with now is to shell into the VM using binary `/Applications/finch/lima/bin/limactl` and run the version command, then compare with `finch version`. This would work in CI

File: cmd/finch/version_test.go
Comment (davidhsingyuchen) [inline cmd/finch/version_test.go:84]: https://github.com/runfinch/finch/pull/83#discussion_r1065119672 is also applicable to this PR.

File: cmd/finch/version.go
Comment (ahsan-z-khan) [reply cmd/finch/version.go:77]: Added  `"Please run finch vm start to start the virtual machine"` when VM is not in a good status

File: cmd/finch/version.go
Comment (ahsan-z-khan) [reply cmd/finch/version.go]: `Fprintf` writes the output to w

File: cmd/finch/version.go
Comment (ahsan-z-khan) [reply cmd/finch/version.go]: updated

File: cmd/finch/version_test.go
Comment (ahsan-z-khan) [reply cmd/finch/version_test.go]: Updated on the latest push. Can you check and suggest how to assert through `&stdOut'

File: cmd/finch/version_test.go
Comment (pendo324) [inline cmd/finch/version_test.go:123]: I don't see `tc.versionOutput` being used anywhere currently. Also, not sure if we want to rely on these exact git commit SHAs in our tests. That would mean we have to update these pretty regularly, right? Maybe for these you can just check that they aren't empty unless there's a way to mock them easily?

Comment (ahsan-z-khan) [reply cmd/finch/version_test.go:123]: These are randomly generated SHA, not the actual ones. But follows the format.

For the structs, I was thinking doing `assert.Contains(t, tc.versionOutput, stdOut.String()` , but errors out. looking into it

File: e2e/version_test.go
Comment (davidhsingyuchen) [reply e2e/version_test.go]: > Other have dependencies on nerdctl output/format.

That's true, but I don't think they will change frequently if ever. At the end of the day, we're printing them to our users, so we may want to verify them.

> 500 characters which fails in the golangci for lll

We can either disable `lll` for that line ([example](https://github.com/runfinch/common-tests/blob/1d5ec0db09b523f47f9825ef7237ed1d9c51401a/tests/run.go#L437)) or [use `(?x)` to span the regex itself across multiple lines](https://superuser.com/a/1748758/730265).

File: cmd/finch/version_test.go
Comment (ahsan-z-khan) [reply cmd/finch/version_test.go]: I’ve been trying to assert through this,
```var output bytes.Buffer
			newVersionAction(lcc, logger, &output).run()
			assert.Contains(t, output.String(), tc.versionOutput.Client)```
But during debug, i can see that `&output`  buffer is not getting filled through the `io.Writer`

File: cmd/finch/version_test.go
Comment (davidhsingyuchen) [inline cmd/finch/version_test.go:101]: We probably want to replace this with a `postRunCheck` ([example](https://github.com/runfinch/finch/blob/52ce4e77aeb679a380f7b580bb13ddd5c8a45b01/pkg/config/nerdctl_config_applier_test.go#L38)). In `postRunCheck`, we can assert each line of the stdout when we expect there's no error and just return when an error is expected (`wantErr` is not `nil`).

```suggestion
		postRunCheck func(t *testing.T, stdout []byte)
```

File: cmd/finch/version.go
Comment (davidhsingyuchen) [inline cmd/finch/version.go:78]: Also, we can print different things depending on the status. If it's nonexistent, then "vm init"; if it's stopped, then "vm start".

```suggestion
		fmt.Println(`Please run "finch vm start" to start the virtual machine to get the version of other components (e.g., nerdctl)`)
```

File: cmd/finch/version.go
Comment (davidhsingyuchen) [inline cmd/finch/version.go:86]: if the error is not nil, we may want to return it instead of a silent fail.

File: e2e/version_test.go
Comment (ningziwen) [inline e2e/version_test.go:25]: Should we inject Version.version here?

Comment (ahsan-z-khan) [reply e2e/version_test.go:25]: Yeah we can do that. 
```
versionInfo := command.StdoutStr(o, "version")
gomega.Expect(versionInfo).Should(gomega.ContainSubstring(version.Version))
gomega.Expect(versionInfo).Should(gomega.ContainSubstring(version.GitCommit))
```

File: cmd/finch/version.go
Comment (ahsan-z-khan) [reply cmd/finch/version.go:78]: I would rather not complicate the command and print different status here, which is already being implement in `finch vm status`. I will just remove this line

File: .golangci.yaml
Comment (davidhsingyuchen) [reply .golangci.yaml:15]: If we cannot break it into multiple lines for some reason, we can add `nolint:lll` and explain the reason behind.

**Comment by ahsan-z-khan** [comment]
Note: Lint passed locally for `nolint` but failing in CI

File: e2e/version_test.go
Comment (ahsan-z-khan) [reply e2e/version_test.go]: added

File: cmd/finch/version_test.go
Comment (pendo324) [reply cmd/finch/version_test.go:85]: I think we're still missing the error test case, but everything else looks good in the latest revision

**Comment by pendo324** [comment]
The linting error looks pretty weird. Let's retry it after the other workflows finish

File: cmd/finch/version_test.go
Comment (ahsan-z-khan) [reply cmd/finch/version_test.go:85]: Not sure why Linter fails in CI during environment preparation. 

Example:
```
  /usr/bin/tar: ../../../go/pkg/mod/inet.af/tcpproxy@v0.0.0-20220326234310-be3ee21c9fa0/.travis.yml: Cannot open: File exists
  Error: /usr/bin/tar: ../../../go/pkg/mod/inet.af/tcpproxy@v0.0.0-20220326234310-be3ee21c9fa0/tcpproxy.go: Cannot open: File exists
```

Comment (pendo324) [reply cmd/finch/version_test.go:85]: Yeah the linter error does look really weird, but I think this comment is about 
having a mock where lima.GetVMStatus returns an error, instead of nil, like

```golang
...
    getVMStatusC.EXPECT().Output().Return(nil, errors.new("Some error"))
...
```

Comment (ahsan-z-khan) [reply cmd/finch/version_test.go:86]: added in new push

File: cmd/finch/version_test.go
Comment (davidhsingyuchen) [inline cmd/finch/version_test.go:20]: To make the name of the testing function align with the name of the tested function.

```suggestion
func TestNewVersionCommand(t *testing.T) {
```

File: cmd/finch/version.go
Comment (davidhsingyuchen) [inline cmd/finch/version.go:82]: - We can use early return here to reduce the nesting levels of the code below.
- We can create a wrapper function that will call `fmt.Fprintf(va.stdOut, "Finch version:\t%s", version.Version)` if the returned error is not nil so that we can remove the duplicated statements (there're 2 more in the code below).
- Return an error when the VM is not running: Checked nerdctl's code, and it also returns an error if some steps fail (after printing whatever is available).

```suggestion
	if err != nil {
		return fmt.Error("failed to get VM status: %w", err)
	}
	if status != lima.Running {
		return errors.New("detailed version info is unavailable because VM is not running")
	}
```

**Comment by ahsan-z-khan** [comment]
> > Note: Lint passed locally for `nolint` but failing in CI
> 
> The linting error looks pretty weird. Let's try retrying it after the other workflows finish

I think I got it. It was `go.sum`. I need to run `go mod tidy' and force pushe

File: cmd/finch/version.go
Comment (davidhsingyuchen) [inline cmd/finch/version.go:95]: Could we prefix the errors?

```suggestion
			return fmt.Errorf("failed to create the nerdctl version command: %w", err)
		}

		var nerdctlVersion NerdctlVersionOutput
		err = json.Unmarshal(out, &nerdctlVersion)
		if err != nil {
			fmt.Fprintf(va.stdOut, "Finch version:\t%s", version.Version)
			return fmt.Errorf("failed to JSON-unmarshal the nerdctl version output: %w", err)
```

File: cmd/finch/version.go
Comment (ahsan-z-khan) [reply cmd/finch/version.go]: updated

File: cmd/finch/version.go
Comment (ahsan-z-khan) [reply cmd/finch/version.go]: updated

**Comment by ahsan-z-khan** [comment]
> > Note: Lint passed locally for `nolint` but failing in CI
> 
> The linting error looks pretty weird. Let's try retrying it after the other workflows finish

It's weird that it was passing before the latest push on `main`. Not sure if that has something to do

File: cmd/finch/version.go
Comment (davidhsingyuchen) [inline cmd/finch/version.go:108]: To make it conform to the thing specified in https://github.com/runfinch/finch/issues/12#issuecomment-1347252973

```suggestion
		fmt.Fprintf(va.stdOut, " nerdctl:\n")
		fmt.Fprintf(va.stdOut, "  Version:\t%s\n", nerdctlVersion.Client.Version)
		fmt.Fprintf(va.stdOut, "  GitCommit:\t%s\n", nerdctlVersion.Client.GitCommit)
		for _, compo := range nerdctlVersion.Client.Components {
			fmt.Fprintf(va.stdOut, " %s:\n", compo.Name)
			fmt.Fprintf(va.stdOut, "  Version:\t%s\n", compo.Version)
			fmt.Fprintf(va.stdOut, "  GitCommit:\t%s\n", compo.Details.GitCommit)
```

File: cmd/finch/version_test.go
Comment (davidhsingyuchen) [inline cmd/finch/version_test.go:105]: nit: prevent panic when there's a length mismatch

```suggestion
require.Len(t, lines, 18)
```

File: cmd/finch/version_test.go
Comment (davidhsingyuchen) [inline cmd/finch/version_test.go:141]: `lines := strings.SplitAfter(string(stdout), "\n")` can be removed.

```suggestion
				assert.Equal(t, string(stdout), "Finch version:\t\n")
```

**Review by pendo324 — DISMISSED** [review]
The implementation and testing LGTM in the latest revision, but I'd prefer to wait for @davidhsingyuchen's review too before merging.

Also, the linter is still failing with some weird errors. I found some resources related to this:

https://github.com/golangci/golangci-lint-action/issues/23
https://github.com/golangci/golangci-lint-action/issues/244

It might just be a rare issue that happened to get triggered this time, but I haven't really looked into it yet

**Review by davidhsingyuchen — COMMENTED** [review]
https://github.com/runfinch/finch/blob/main/CONTRIBUTING.md#pull-request-lifecycle: Please avoid resolving the comments left by a PR reviewer yourself, thanks!

**Comment by ahsan-z-khan** [comment]
> > Note: Lint passed locally for `nolint` but failing in CI
> 
> The linting error looks pretty weird. Let's try retrying it after the other workflows finish

According to this: https://github.com/golangci/golangci-lint-action/issues/23

Setting 'skip-pkg-cache: true' on each run of the golangci-lint action seems to fix it. We have it setup as 'false'. Another solution is to config the ` go-linter`  in a separate action.

**Comment by davidhsingyuchen** [comment]
> > > Note: Lint passed locally for `nolint` but failing in CI
> > 
> > 
> > The linting error looks pretty weird. Let's try retrying it after the other workflows finish
> 
> According to this: [golangci/golangci-lint-action#23](https://github.com/golangci/golangci-lint-action/issues/23)
> 
> Setting 'skip-pkg-cache: true' on each run of the golangci-lint action seems to fix it. We have it setup as 'false'. Another solution is to config the ` go-linter` in a separate action.

If we run `make lint` locally, some fixes are triggered, which means that it won't pass in CI. That line will be removed after addressing https://github.com/runfinch/finch/pull/108#discussion_r1066452161 anyway.

```sh
➜  finch git:(version_extension) make lint
golangci-lint run
➜  finch git:(version_extension) ✗ git --no-pager diff
diff --git a/cmd/finch/main.go b/cmd/finch/main.go
index 5b0f274..83c6889 100644
--- a/cmd/finch/main.go
+++ b/cmd/finch/main.go
@@ -39,7 +39,7 @@ func main() {
 	}
 }

-//nolint: lll //Line is more than 150 character
+// nolint: lll //Line is more than 150 character
 func xmain(logger flog.Logger, ffd path.FinchFinderDeps, fs afero.Fs, loadCfgDeps config.LoadSystemDeps, mem fmemory.Memory, stdOut io.Writer) error {
 	fp, err := path.FindFinch(ffd)
 	if err != nil {
```

File: cmd/finch/version_test.go
Comment (davidhsingyuchen) [inline cmd/finch/version_test.go:126]: nit, ensure that nothing else is on that line

```suggestion
				assert.Equal(t, lines[17], "  GitCommit:\tv1.1.4-0-g5fd4c4d1\n")
				assert.Equal(t, lines[18], "")
```

File: cmd/finch/version.go
Comment (davidhsingyuchen) [reply cmd/finch/version.go]: My bad, I meant the following:

```go
func (va *versionAction) run() error {
	if err := va.printVersion(); err != nil {
		// Now we only need this statement once, which is clearer and easier to maintain.
		fmt.Fprintf(va.stdOut, "Finch version:\t%s\n", version.Version)
		return err
	}
	return nil
}

func (va *versionAction) printVersion() error {
	// Move the main logic to this function
	// ...
}
```

**Review by davidhsingyuchen — DISMISSED** [review]
Feel free to do the suggested changes in a follow-up PR.

File: cmd/finch/version_test.go
Comment (ahsan-z-khan) [reply cmd/finch/version_test.go:126]: okay sure. lets merge this and get the release ready. in the meantime I'll open follow up PR once releases

File: cmd/finch/version.go
Comment (davidhsingyuchen) [reply cmd/finch/version.go]: Please create a follow-up PR for this. Resolving this now.

**Review by davidhsingyuchen — APPROVED** [review]

**Comment by ahsan-z-khan** [comment]
6 `e2e` tests are failing on `arm64` which are not related to the PR. They demonstrate the behavior of Issue #140 .

File: cmd/finch/version.go
Comment (davidhsingyuchen) [reply cmd/finch/version.go]: https://github.com/runfinch/finch/pull/147

**Pull request merged**
